### PR TITLE
Make first order gradient graphs more efficient

### PR DESCRIPTION
**Pull request by t-vi**
Previously, nodes are visited as often as they are used and each time a
derivative is computed. Only at the leaves were the contributions of
everything added. This patch changes this to add at any node that is
used several times.

### CODE REVIEW

**Comment by t-vi** [comment]
This pull request is only about the second commit, the first is #5946 .
I noticed that my gradient had many more O^3 (matmul etc.) operations than it should have and tracked this down to how gradients are computed when a value is used several times in the computation.
Graphs are becoming really big and unwieldy if they are not purely sequential computation. Also, the duplication cannot be eliminated by CSE because the "output part" is duplicate rather than the input (one could, in theory commute add with all the gradient ops).
While it doesn't fix anything, it might also have a mitigating impact for people seeing other effects when working with first order gradients (e.g. #4534).

File: tests/python/relay/test_pass_gradient.py [hunk 1]
```diff
@@ -27,6 +29,20 @@
 import tvm.relay.op as op
 
 
+def count_ops(expr):
```
Comment (MarisaKirisame) [inline tests/python/relay/test_pass_gradient.py:32]: can you move this to some common file? this look like something useful to other places as well.

**Comment by MarisaKirisame** [comment]
We was thinking of using ANF then AD. But this also work.

File: tests/python/relay/test_pass_gradient.py (hunk 1 above)
Comment (t-vi) [reply tests/python/relay/test_pass_gradient.py:32]: I'll move it to `python/tvm/relay/testing/__init__.py` when rebasing post #5946 .

**Comment by MarisaKirisame** [comment]
@t-vi please rebase.

**Comment by t-vi** [comment]
@MarisaKirisame  Thank you! I rebased and now the CI is all happy again.

**Comment by tqchen** [comment]
Thanks @t-vi @MarisaKirisame

**Pull request merged**